client/network/common/src/header_backend.rs

```diff
@@ -0,0 +1,65 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use sp_blockchain::{HeaderBackend, Info, Result};
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, Header as HeaderT, NumberFor},
+};
+
+/// Wrapper trait for [`HeaderBackend`](sp_blockchain::HeaderBackend) narrower parameter.
```

Comment:
1. I think you've accidentally dropped a word from this sentence? (:
2. As far as I can see this trait is also missing the `status` method and bunch of implicit methods which `HeaderBackend` has, so wouldn't it be more appropriate to say that it's a cut down, more narrowly-scoped variant of that trait, or something along those lines?
3. If it's a more narrowly scoped variant of `HeaderBackend` wouldn't it maybe make sense to make it a supertrait of that one? (Or not; I don't know.)
4. Hm, is this trait actually necessary? Like, for example, we have this method in the original trait:

```rust
fn header(&self, id: BlockId<Block>) -> Result<Option<Block::Header>>;
```

...so can't its implementation just be something like this for the case where we internally only want to deal with a `Hash` as the parameter? Since the trait *does* provide the `hash` method to convert a block number into a block hash anyway.

```rust
fn header(&self, id: BlockId<Block>) -> Result<Option<Block::Header>> {
    match id {
        BlockId::Hash(hash) => self.header_impl(hash),
        BlockId::Number(number) => {
            if let Some(hash) = self.hash(number)? {
                self.header_impl(hash)
            } else {
                Ok(None)
            }
        }
    }
}
```

...and then once we remove the block numbers from the API this could be removed and then the original trait would be essentially the same as what this new trait is? (except the `status` method, I guess)

Reply:
Indeed we could get away without this trait. However, transforming the Number to a hash requires an extra RPC call in cumulus later, which I wanted to avoid. This trait makes sure that no one passes in a number. 

Maybe it's overengineered since it seems unlikely that a block number is passed here. So I dont have a strong opinion on it and could remove if you think its unnecessary.

Reply:
> However, transforming the Number to a hash requires an extra RPC call in cumulus later, which I wanted to avoid. This trait makes sure that no one passes in a number.

Even if someone does pass a number is this going to actually make a difference? Assuming there's no actual functional difference the way I see it there are two possibilities; either:

1) the RPC calls are cheap enough that this is not going to make any difference at all even if someone calls it,
2) or the RPC calls are relatively expensive so it is going to make a difference, in which case we should fix them so that they are cheap enough.

I guess it could make a difference if someone's connecting to an RPC node over a high latency link on another machine? But we don't really want to support such configurations anyway. (And of course this is assuming that these will get called with a block number.)

Personally I think I'd lean on getting rid of extra trait (we already have a lot of them), but it's not the end of the world if it stays this way. Maybe another reviewer could chime in whichever they prefer and then we can go with that?

Reply:
I think you are right, let me remove it.